File: data_structures/javascript/singlyLinkedList.js
//Value of Linked List
//No need for a fixed allocation of memory in a defined space, because the nodes can be scattered in memory dynamically

//When would we want to use this?
//If we don't know how much space we'll need, this structure makes more sense than an array, which requires that we allocate a fixed amount of memory
//Linked lists have no added memory for storage, but they do have added memory for the pointers to the nodes in the array (this is less than the added memory storage allocation, in general)
//Adding an element to the head of the list is very easy, whereas with an array, it requires that we shift the position of all the subsequent elements
//Adding an element to the end of an array is an O(1) operation if we're using a dynamic ring buffer that tracks the current last-filled element, but O(current number of elements) if we do not have that ring buffer. Similarly, we have O(current number of elements) to insert into a singly linked list, and O(1) to insert at the end of an doubly-linked list
//For insertion, different variations of these data structuers have various advantages, making them score roughly evenly.

//When would we not want to use this? 
//If we want constant time look up, we should use an array instead. Because arrays have a fixed starting memory address, they can be randomly accessed in O(1) time because the address of a locations are mathematically fixed in relation to this starting memory address.

class SinglyLinkedNode {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class SinglyLinkedList {
    constructor() {
        this.length = 0;
        this.head = null;
    }

    //O(this.length)
    add(data) {
        const newNode = new SinglyLinkedNode(data);

        let currentNode = this.head;

        if (!currentNode) {
            this.head = newNode;
            this.length++;
            return newNode;
        } else {
            while (currentNode.next) {
                currentNode = currentNode.next;
            }
            currentNode.next = newNode;
            this.length++;
            return newNode;
        }
    }

    //O(position)
    findNodeAtPosition(position) {
        let currentNode = this.head;
        let currentPosition = 1;

        if (position > this.length || position < 1 || this.length === 0) {
            throw new RangeError(`Position ${position} is not available in LinkedList of length ${this.length}`);
        }

        while (currentPosition < position) {
            currentNode = currentNode.next;
            currentPosition++;
        }

        return currentNode;
    }

    //O(this.length)
    findNodeWithData(data) {
        let currentNode = this.head;

        if (!currentNode) {
            return null;
        }

        while (currentNode.next && currentNode.data != data) {
            currentNode = currentNode.next;
        }
        if (currentNode.data == data) {
            return currentNode;
        } else {
            return null;
        }
    }

    //O(this.length)
    findNodeWithDataStrict(data) {
        let currentNode = this.head;

        if (!currentNode) {
            return null;
        }

        while (currentNode.next && currentNode.data !== data) {
            currentNode = currentNode.next;
        }
        if (currentNode.data === data) {
            return currentNode;
        } else {
            return null;
        }
    }

    //O(position)
    removeNodeAtPosition(position) {
        let currentNode = this.head;
        let currentPosition = 0;
        let previousNode = null;
        let NodeToDelete = null;
        let deletedNode = null;

        if (position > this.length || position < 1 || this.length === 0) {
            throw new RangeError(`Position ${position} is not available in LinkedList of length ${this.length}`);
        }

        //remove the head
        if (position === 1) {
            this.head = currentNode.next;
            deletedNode = currentNode;
            currentNode = null;
            this.length--;
            return deletedNode;
        } else {
            //start position at 0 so that we break when NodeToDelete is currentNode.next
            //and not currentNode
            while (currentPosition < position) {
                previousNode = currentNode;
                NodeToDelete = currentNode.next;
                currentPosition++;
            }

            previousNode.next = NodeToDelete.next;
            deletedNode = NodeToDelete;
            //set to null to formally remove from linkedList
            NodeToDelete = null;
            this.length--;
            return deletedNode;
        }
    }

    //O(this.length)
    removeNodeWithData(data) {
        let currentNode = this.head;
        let deletedNode = null;
        let previousNode = null;

        if (!currentNode) {
            return null;
        }

        while (currentNode.next && currentNode.next.data != data) {
            previousNode = currentNode;
            currentNode = currentNode.next;
        }

        if (currentNode.data == data) {
            previousNode.next = currentNode.next;
            deletedNode = currentNode;
            currentNode = null;
            this.length--;
            return deletedNode;
        } else {
            return null;
        }
    }

    //O(this.length)
    removeNodeWithDataStrict(data) {
        let currentNode = this.head;
        let deletedNode = null;
        let previousNode = null;

        if (!currentNode) {
            return null;
        }

        while (currentNode.next && currentNode.next.data !== data) {
            previousNode = currentNode;
            currentNode = currentNode.next;
        }

        if (currentNode.data === data) {
            previousNode.next = currentNode.next;
            deletedNode = currentNode;
            currentNode = null;
            this.length--;
            return deletedNode;
        } else {
            return null;
        }
    }
}

File: data_structures/ruby/singly_linked_list.rb
class SinglyLinkedListNode
    attr_accessor :data, :next
    def initialize(data)
        @data = data
        @next = nil
    end
end

class SinglyLinkedList
    def initialize
        @length = 0;
        @head = nil
    end

    def add(data)
        new_node = SinglyLinkedListNode.new(data)

        current_node = @head

        if @head.nil?
            @head = new_node
        else
            until current_node.next.nil?
                current_node = current_node.next
            end
            current_node.next = new_node
        end

        @length += 1
        return new_node
    end

    def find_node_at_position(position)
        current_node = @head
        current_position = 1

        if position > @length || position < 1
            raise ArgumentError, "Position must be within range(0, #{@length}) or list is currently empty"
        end

        until current_position == position
            current_node = current_node.next
            current_position += 1
        end

        return current_node
    end

    def find_node_with_data(data)
        current_node = @head

        if current_node.nil?
            return nil
        else
            until current_node.next.nil || current_node.data == data
                current_node = current_node.next
            end
            return current_node.data == data ? current_node : nil
        end
    end

    def remove_node_at_position(position)
        current_node = @head
        current_position = 0
        previous_node = nil
        node_to_delete = nil
        deleted_node = nil

        if position > @length || position < 1
            raise ArgumentError, "Position must be within range(0, #{@length})"
        end

        if (position == 1)
            @head = current_node.next
            deleted_node = current_node
            current_node = nil
        else
            until current_position == position
                previous_node = current_node
                node_to_delete = current_node.next
                current_position += 1
            end
            previous_node.next = node_to_delete.next 
            deleted_node = node_to_delete
            node_to_delete = nil
        end

        @length -= 1
        return deleted_node
    end

    def remove_node_with_data(data)
        current_node = @head
        deleted_node = nil
        previous_node = nil

        if current_node.nil?
            return nil
        elsif (@head.data == data)
            deleted_node = @head
            @head = @head.next
            return deleted_node
        else
            until current_node.next.nil? || current_node.data == data
                previous_node = current_node
                current_node = current_node.next
            end
            if (current_node.data == data)
                previous_node.next = current_node.next
                deleted_node = current_node
                current_node - nil
                @length -= 1
                return deleted_node
            else
                return nil
            end
        end
    end

end